runtime/onert/core/src/backend/cpu_common/MemoryPlanner.cc

```diff
@@ -27,8 +27,6 @@ namespace cpu_common
 
 void BumpPlanner::claim(const ir::OperandIndex &ind, size_t size)
 {
-  assert(size != 0);
-
```

Comment:
@wateret Is it safe to just remove `assert`? What happens when `size=0`? It may cause `malloc(0)`, It would be good if `size=0` case is handled explicitly.

Reply:
The memory planners never return null. They have one big memory chunk and the actual buffer address is calculated `base + offset`.

However I agree that we need some exact policy for 0-sized tensors in general(not only these static planners but also dyanmic allocators). We need to define what to handle 0-sized tensors. #4744 has a different approach from this PR, that always return null for 0-sized tensors.

Reply:
However I ended up with allowing 0-sized tensor allocation since I think we may have to distinguish these two:

- optional tensor which were not given  (In python, `None`)
- a tensor with no values (In python, `[]`)

And this case is the second, but with the way from #4744, we won't be able to distinguish those.

Reply:
> The memory planners never return null. They have one big memory chunk and the actual buffer address is calculated `base + offset`.
> 

@wateret Can tensor A and B have same address if tensor A is allocated then next tensor B (with size 0) is allocated? If some implementation (like cache) would like to use the address as an unique key, it may break the assumption.

Reply:
> Can tensor A and B have same address if tensor A is allocated then next tensor B (with size 0) is allocated?

Yes.

> If some implementation (like cache) would like to use the address as an unique key, it may break the assumption.

Good point. But would that actually break that caching assumption? I took a look at 

https://github.com/Samsung/ONE/blob/083789b16ef07d6e6ab14134db6537077fe7e76f/runtime/onert/backend/cpu/ops/FullyConnectedLayer.cc#L244-L251

but I don't see it is an issue that a tensor buffer is the same address with another tensor buffer. Maybe there is another case that would be an issue? Please correct me if I'm wrong.

Reply:
@wateret For `FullyConnectedHybrid()`, I am not sure because I don't know the details how ruy handles the cache keys. Furthermore, I don't know the all cases in `onert`. Thus, I am not 100% sure this change is safe.

Before this change, it was checked with `assert`. It should be very rare case. I would choose to handle this case in early stage if possible. `[0]` is used for `unspecified` dimension, however, I think it is specified during running by static or dynamic shape inferer.

Reply:
> `[0]` is used for `unspecified` dimension, however, I think it is specified during running by static or dynamic shape inferer.

You're right. So this should be more about nnapi frontend than allocating 0.

---

Anyways now that I got interested in 0-size tensors so I will keep working on this too. I think we could make some cases when there are 0-size tensors intentionally. I'll dig into it.

> @wateret For `FullyConnectedHybrid()`, I am not sure because I don't know the details how ruy handles the cache keys. Furthermore, I don't know the all cases in `onert`. Thus, I am not 100% sure this change is safe.

About caching, I still think there is no issue, so I am going to keep this way(two tensors could have a same buffer pointer). Once we find that could restrict use of ruy or other kernels then we can restrict it - "the returned tensor pointers must be distinct for the all tensors that interfere their lifetimes"

Reply:
@glistening I had second thoughts that we may still need to allow this. Is there a way to distinguish `[0]` is a dynamic tensor or actually a 0-size tensor in NN API?

- According to https://github.com/Samsung/ONE/issues/4793#issuecomment-717086013 it may mean it is always a dynamic tensor. (Then would that mean NN API can't express 0-size tensor?) Then as we assume the unspecified dimension value to 1 in tflite, we can go this way.
- If both are possible, we should allocate 0-size tensor in the static tensor memory area, and if we don't change anything in our allocators, it may return the same pointer for different tensors, but as I mentioned above, I don't think that is an issue.